Update docs related to Workspace tab (was "Filesystem" tab) (#3432)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/workspaces/index.md

Author: in pngs, show /Demos/workspace/ dir instead of /app/ dir, in Workspace tab of Sources tool.

Author: Make article a simple tut flow:
* Remove some sections, remove some pngs, simplify headings.
   * Change Address bar to localhost url (on a couple remaining pngs).  Might just remove those pngs.
   * Show Workspace tab not Page tab, of Sources tool (on a couple remaining pngs).  Might just remove those pngs.

DevTools doesn't support every framework variation.  If you run into issues while using workspaces with your framework of choice, or you identify framework-specific steps that are needed, please reach out to use by opening an issue on the [MicrosoftEdge/DevTools](https://github.com/MicrosoftEdge/DevTools/issues) repository.

<!-- ====================================================================== -->
## Editing CSS by using the Styles tab in the Elements tool, when using the Workspace tab of the Sources tool

If you use a workspace, in some cases you can edit CSS in the **Styles** tab in the **Elements** tool and have the changes saved to the mapped file on disk:

* Edits that you make in the **Styles** tab in the **Elements** tool _are_ saved to the CSS file on disk, if DevTools has mapped a style sheet on the webpage and the file in the workspace, such as by a source map, or by matching content.

>    If you use an automatic build system framework, changes are saved to your source file if DEvTools was able to map the style sheet to the workspace file, by using a source map.

Reviewer: nit: DevTools is spelled DEvTools

                                                  

<!-- ------------------------------ -->
#### Related feature: Local Overrides
<!-- https://developer.chrome.com/docs/devtools/workspaces#overrides -->

**Local overrides** is a DevTools feature that's similar to a workspace.  You can use an override when you want to experiment with changes to a webpage, and you need to display the changes across webpage loads, but you don't care about mapping your changes to the source code of the webpage.

**Local Overrides** saves the changes you make to a webpage in a folder of your choice, and organizes them in a folder structure that matches the URL of overridden resources.

The **Overrides** feature stores a copy of the webpage resources that you edit in DevTools.  When you refresh the page, Microsoft Edge loads the local modified copy of files instead of the files on the server.

See also:
* [Override webpage resources with local copies (Overrides tab)](../javascript/overrides.md)

<!-- ====================================================================== -->

> ## Step 1: Setup

Reviewer 2: "Setup" is kind of short and abstract. What about "Setup the code, server, and workspace"?

Author: setup is noun
set up is verb
result:
```
## Step 1: Set up the code, server, and workspace
```
This fix works well here.

@captainbrosset 

Given that In this article nav shows h2s only, it's reasonable to consider the strategy of forcing all headings to h2, then demoting only the lowest-value headings. 

Also remember option of adding detailed TOC at top, to consider for these several articles.

   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.  -->
# Edit and save files in a workspace
<!-- https://developer.chrome.com/docs/devtools/workspaces -->

> Use the **Workspace** tab in the **Sources** tool to define a workspace, to save DevTools changes in your source code files rather than only in a transient copy of the files that are returned by the web server.

Reviewer 2: ```suggestion
Use the **Workspace** tab in the **Sources** tool to display and edit the contents of a folder on your computer from within the **Sources** tool.  The changes that you make to the files in a workspace are saved to your computer.

The typical way to set up a workspace is to set it up to display the source files of your website's frontend code. This way, you use DevTools both to inspect your website, and to make changes to its source files, and the changes that you make are saved on your computer, rather than lost when you refresh the page.
```

Author: ```
Use the **Workspace** tab in the **Sources** tool to display and edit the contents of a folder on your computer from within the **Sources** tool.  The changes that you make to the files in a workspace are saved to your computer.

The typical way to set up a workspace is to set it up to display the source files of your website's frontend code.  This way, you use DevTools both to inspect your website, and to make changes to its source files.  This way, the changes that you make are saved on your computer, rather than lost when you refresh the page.
```

> This tutorial provides hands-on practice in setting up and using a workspace in DevTools.  After you add files to a workspace, changes that you make in your source code by using DevTools are saved on your local computer.  These changes are preserved across page refreshes.

Reviewer 2: ```suggestion
This tutorial provides hands-on practice in setting up and using a workspace in DevTools.  After you set up to a workspace, the changes that you make to the files in the workspace, by using DevTools, are saved on your local computer.
```

Author: omit xtra to

See also:

> * [Getting started with the web](https://developer.mozilla.org/docs/Learn/Getting_started_with_the_web) - How to use HTML, CSS, and JavaScript to build a web page. 

Reviewer 2: Not sure why we link to Getting started with the web here. Knowing what the web is, and knowing about HTML, CSS, and JS is kind of a prereq for the entire DevTools, not just the Workspace tab. So, either we tell that to readers in our overview article only, or we mention it everywhere. But mentioning it only here feels odd to me.

Author: inspecting doc design: upstream vs. downstream PR vs. what it should be:
upstream: https://developer.chrome.com/docs/devtools/workspaces
starts w/ 
![image](https://github.com/user-attachments/assets/db20bc2d-8775-40c3-944d-b364c21d7c93)
* https://developer.mozilla.org/docs/Learn/Getting_started_with_the_web (link text mismatch/outdated, latest dest title is "Your first website")
* https://developer.chrome.com/docs/devtools/css
* https://developer.mozilla.org/docs/Learn/Common_questions/set_up_a_local_testing_server#Running_a_simple_local_HTTP_server

@captainbrosset

Author: since I have made this Workspace tut standalone and targeted all user levels (Maint'ers) w/ end-to-end steps spelled out, that reduces any need to link-away.  i'll at least move the 3 links to See Also at bottom, or move them to an Overview article.

* [Get started viewing and changing CSS](../css/index.md) - How to use DevTools to make basic changes to CSS.
* [Set up a localhost server](../../visual-studio-code/microsoft-edge-devtools-extension/install.md#step-6-set-up-a-localhost-server) in _Installing the DevTools extension for Visual Studio Code_.  How to run a local HTTP web server by using Node.js.
* [Running a simple local HTTP server](https://developer.mozilla.org/docs/Learn/Common_questions/set_up_a_local_testing_server#running_a_simple_local_http_server) in _How do you set up a local testing server?_ - Using Python.

<!-- ====================================================================== -->
## Overview
<!-- https://developer.chrome.com/docs/devtools/workspaces#overview -->

> A DevTools _workspace_ lets you save changes that you make to a local copy of the source code to the same file on your computer, so that changes are retained across refreshes of the page.  Here's a typical scenario for using a workspace:

Reviewer 2: A "local copy" of the source code is confusing to me. The Workspace tab doesn't copy the code. It shows the actual files on disk (just like a text editor would do).

```suggestion
A DevTools _workspace_ lets you view your website's source code that's on your computer, and make changes to it, from the **Sources** tool, so that changes are retained across refreshes of the page.  Here's a typical scenario for using a workspace:
```

> * You have the source code for the demo website on your desktop.

Reviewer 2: ```suggestion
* You have the source code for your website on your computer.
```

> * You are running a local web server from the source code directory, so that the site is accessible at `localhost:8080`.  Note: If you use the Python server option, the default port number is `8000`.

Reviewer 2: ```suggestion
* You are running a local web server from the source code directory, so that the site is accessible at a local development URL such as `http://localhost:8080`.
```

> * You opened `localhost:8080` in Microsoft Edge, and you are using DevTools to change the website source code which includes the CSS, HTML, and JavaScript files. 

Reviewer 2: ```suggestion
* You open `http://localhost:8080` in Microsoft Edge, and use the **Sources** tool in DevTools to change the source code of your website, including the CSS, HTML, and JavaScript files.

* You reload the page to see the changes.

  Your website might use a build system which you need to run before you can see the changes. Developers often use build systems which watch for changes in the source code files to build those changes and then automatically refresh the page in the browser. If that's your case, making a change to a source file in the **Sources** tool is the only thing you need to do to see those changes applied to the rendered web page.
```

It's important to mention the watch-build-reload system that a lot of devs use nowadays, because that's what makes the Workspace tab really useful. If you don't use this kind of system, that means you need to reload the page manually after each change, which is not much more useful than just editing the files in your text editor.

Author: ```
* You open `http://localhost:8080` in Microsoft Edge, and use the **Sources** tool in 
DevTools to change the source code of your website, including the CSS, HTML, and JavaScript files.

* You reload the page to see the changes.

  Your website might use a build system which you need to run before you can see the 
changes.  Web developers often use a build system that watches for saved changes in the 
source code files.  When you save a change to a source code file, the build system builds those 
changes and then automatically refreshes the page in the browser.

  If you use such an automated build system, when you make a change to a source file in the **Sources** 
tool and save it, you automatically immediately see those changes applied to the rendered web page.
```

1. Go to [MicrosoftEdge / Demos](https://github.com/MicrosoftEdge/Demos) in a new window or tab.

1. Click the **Code** drop-down button, and then click the **Copy url to clipboard** button.

   The URL is copied to the clipboard: `https://github.com/MicrosoftEdge/Demos.git`

   Or, if you have GitHub Desktop installed, click **Open with GitHub Desktop** to clone the repo, and skip the command prompt step below.

   Or, you can use Visual Studio Code's **Source Control** pane to clone the repo, and skip the command prompt step below.

1. Assuming you have git installed, open a command prompt.

1. Clone the repo to your local drive, entering the URL string that you copied from the GitHub repo.  If you use a command prompt:

   ```Shell
   # example location where the repo directory will be added:
   cd c:/users/localAccount/github/
   git clone https://github.com/MicrosoftEdge/Demos.git
   ```

For details about cloning a repo, see:
* [Cloning a repository](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) - GitHub docs.

Continue with the next section.

<!-- ---------- -->
###### Create a directory for the localhost server to use

> 1. Create an `app` directory, such as `~/Desktop/app` or `C:\Users\localAccount\app\`.

Author: this isn't necessary.  if using the cloned Demos repo, it makes more sense to add the entire cloned /Demos/ dir or the cloned /Demos/demo-to-do/ dir to the Workspace (can switch to a test branch, not "main") instead of needless copying from /Demos/demo-to-do/ dir to a separate /app/ dir.

If we change this approach to skip creating /app/ dir, would need to update text & pngs.
@captainbrosset

Author: updated text.  entered fresh Comment to update pngs to not show /app/ dir.

<!-- ------------------------------ -->
#### Limitations
<!-- https://developer.chrome.com/docs/devtools/workspaces#limitations -->

If you're using a modern framework, it might transform your source code from a format that's easy to maintain into a format that's optimized to run as quickly as possible.  A workspace is usually able to map the optimized code back to the original source code, by using [source maps](https://blog.teamtreehouse.com/introduction-source-maps) for JavaScript and CSS.  However, there's a lot of variation in how each framework uses source maps.

DevTools doesn't support every framework variation; for example, the Workspaces feature (**Workspace** tab) doesn't work with the Create React App framework.

> If you run into issues while using workspaces with your framework of choice, or you identify framework-specific steps that are needed, [start a thread in the Chrome DevTools mailing list](https://groups.google.com/forum/#!forum/google-chrome-developer-tools) or [ask a question on Stack Overflow](https://stackoverflow.com/questions/ask?tags=google-chrome-devtools) to exchange information with the rest of the DevTools community.

Reviewer 2: ```suggestion
DevTools doesn't support every framework variation. If you run into issues while using workspaces with your framework of choice, or you identify framework-specific steps that are needed, please reach out to use by opening an issue on the [MicrosoftEdge/DevTools](https://github.com/MicrosoftEdge/DevTools/issues) repository.
```

                                                                     

<!-- ------------------------------ -->
#### Related feature: Local Overrides
<!-- https://developer.chrome.com/docs/devtools/workspaces#overrides -->

> **Local overrides** is a DevTools feature that's similar to a workspace.  You can use an override when you want to experiment with changes to a webpage, and you need to display the changes across webpage loads, but you don't care about mapping your changes to the source code of the webpage.  However, your changes aren't saved when you refresh the webpage. 

Reviewer 2: ```suggestion
**Local overrides** is a DevTools feature that's similar to a workspace.  You can use an override when you want to experiment with changes to a webpage, and you need to display the changes across webpage loads, but you don't care about mapping your changes to the source code of the webpage.  **Local Overrides** saves the changes you make to a webpage in a folder of your choice and organizes them in a folder structure which matches the URL of overridden resources.
```

The last sentence was confusing. The overrides you make _are_ saved across reloads. But it's important to mention that you're not supposed to use this feature for mapping to your source files. When you set up a folder for local overrides the first time, that folder should be empty. DevTools will create all of the folders and sub-folders it needs to store overrides within it.

> The **Overrides** feature lets you store a local copy of the webpage files on the server. When you refresh the page, Microsoft Edge loads the local copy of files instead of the files on the server.

Reviewer 2: ```suggestion
The **Overrides** feature stores a copy of the webpage resources that you edit in DevTools. When you refresh the page, Microsoft Edge loads the local modified copy of files instead of the files on the server.
```

   In the **Workspace** tab are page icons that have a green dot, for `index.html`, `script.js`, and `styles.css`. <!--(The two-way arrow colors are mapped to `.html`, `.js`, and `.css` file types.)-->  The green dot indicates that DevTools has established a mapping between a network resource of the page that's received from the web server, and the local source file in your `/Demos/workspace/` directory:

   ![The Workspace tab has a green dot indicating a mapping between a resource received from the server and a local source file](./index-images/sources-workspace-folder.png)

Continue with the next section.

<!-- ====================================================================== -->
## Step 2: Save a CSS change to disk
<!-- https://developer.chrome.com/docs/devtools/workspaces/#css -->

Next, you edit CSS and save changes to the source file.

To make a change in the CSS file and save it to disk:

1. In the **Sources** tool, in the **Workspace** tab (grouped with the **Page** tab), select `styles.css` to open it in the editor pane.  The `color` property of the `h1` element is set to `fuchsia`:

   ![View styles.css in a text editor](./index-images/sources-workspace-css.png)

> 1. Select the **Elements** (![Elements tool icon](./index-images/elements-tool-icon.png)) tool, and then in the DOM tree, expand the `<body>` element, and then select the `<h1>` element.

Reviewer 2: It's worth adding a paragraph somewhere around here about using the Elements tool, and how it might not always work for everyone.

Imagine a project that uses a framework or pre-processor of some kind to write CSS, whereby developers don't actually write CSS, but a different version of it, which maybe uses a different language, and different file structure.
The project might use a build system which then translates/compiles/combines the CSS into something the browser can understand.
In many cases, the project would also generate source maps, so that DevTools can still allow developers to see and edit their source files in the Elements tool. But in some cases, that might not work. Or the project might not be using source maps at all.
When that happens, the changes made in the Elements tool will _not_ be saved to the workspace files, and we should advise users to edit the source files in the Sources tool instead.
In that imaginary scenario, the files that use a different language would be visible in the Workspace tab, and developers would be able to edit them there. Then, behind the scenes, the changes would need to be re-built/compiled (which may happen automatically), and then the changes would be visible in the rendered web page after a page refresh.

So, mentioning this here would make sense.

Author: reluctant to add headings or pseudo headings.  result:
```
Next, you edit CSS and save changes to the source file.

Using the **Elements** might not always work for everyone.  Suppose your project uses a framework or pre-processor to write CSS, where you don't directly write CSS, but you write the equivalent of CSS by using a different language and a different file structure.  Such a project might use a build system which then translates, compiles, and combines the CSS-like code into something that the browser can understand.

Such a project might generate source maps, so that DevTools allows you to see and edit your source files in the **Elements** tool.  But in some cases, that might not work, or the project might not use source maps.

When that happens, any changes you make in the **Elements** are not saved to the workspace files.  In that case, edit your source files in the **Sources** tool instead.  In the above scenario, the files that use a different language are visible in the **Workspace** tab of the **Sources** tool, and you can edit the source files there.

Then, behind the scenes, any saved changes need to be re-built and compiled (which may happen automatically), and then the changes are visible in the rendered web page after a page refresh (which may happen automatically).

To make a change in the CSS file and save it to disk:
```

   File Explorer or Finder opens.

1. Open the copy of `styles.css` that's in your `/Demos/workspace/` directory in a text editor, such as Visual Studio Code.  The `color` property is now set to the new color, which is orange in this example.  The change was not only made in the copy of the file returned from the web server; the change was also made in your mapped file in your `/Demos/workspace/` workspace directory.

1. Refresh the page.

The color of the `<h1>` element is still set to the new color.  The change remains across a refresh, because when you made the change, DevTools saved the change to disk.  When you refreshed the page, your local server served the modified copy of the file from disk.

**Tip:** You can also change the color by clicking the fucshia-colored swatch to open the color picker to pick a new color. The HEX value for the color you pick is the color name.

Continue with the next section.

<!-- ====================================================================== -->
## Step 3: Save an HTML change to disk
<!-- https://developer.chrome.com/docs/devtools/workspaces/#html -->

Next, edit HTML and save changes to the source file.

> In the **Elements** tool, it's possible to change HTML tagging in a copy of the file that's returned by the server.  However, to save your edits to a local source file, you need to use the **Sources** tool instead of the **Elements** tool.

Reviewer 2: ```suggestion
In the **Elements** tool, it's possible to change the DOM representation of the HTML document, which the browser engine constructed when it loaded the page.  However, changes to the DOM can't be mapped back to HTML source files.  To make changes to an HTML source file, you need to use the **Sources** tool instead of the **Elements** tool.
```

Continue with the next section.

<!-- ---------- -->
###### Optional: Why it doesn't work
<!-- https://developer.chrome.com/docs/devtools/workspaces#why -->

This section describes why the workflow from the above section "Changing the DOM tree in the Elements tool doesn't save changes" doesn't work.

* The tree of nodes that you see on the **Elements** tool represents the page's DOM.

* To display a page, a browser fetches HTML over the network, parses the HTML, and then converts it into a tree of DOM nodes.

* If the page has any JavaScript, that JavaScript may add, delete, or change DOM nodes. CSS can change the DOM, too, via the `content` property.

* The browser eventually uses the DOM to determine what content it should present to browser users.

* Therefore, the final state of the page that users see may be very different from the HTML that the browser fetched.

* This makes it difficult for DevTools to resolve where a change made in the **Elements** tool should be saved, because the DOM is affected by HTML, JavaScript, and CSS.

> In short, the **DOM Tree** !== HTML.

Reviewer 2: ```suggestion
In short, the DOM tree, which the browser engine constructed, is different from the HTML document that was downloaded from the server.
```

   The **index.html** tab opens in the **Sources** tool.

1. In the the `<h1>` tag, replace "DevTools Workspaces Demo" with "I Love Cake".

   An asterisk appears next to `index.html`.

1. Press **Ctrl+S** (Windows, Linux) or **Command+S** (macOS).

   Your change is saved, and the asterisk disappears.

   You made the change in the file listing, rather than in the DOM tree in the **Elements** tool.

1. Refresh the page.  The heading in the rendered page changes to "I Love Cake", because that string was saved in `index.html` in your mapped `/Demos/workspace/` directory:

   ![Change HTML from the Sources tool](./index-images/sources-page-h1.png)

1. In the **Sources** tool, right-click `index.html`, and then select **Open in containing folder**.

   File Explorer or Finder opens.

1. Open the copy of `index.html` that's in your `/Demos/workspace/` directory in a text editor, such as Visual Studio Code.

   The `<h1>` element contains the new text, "I Love Cake", because you made the change using the **Sources** tool's editor to edit `index.html` and then saved the change, and that file was mapped in a workspace (the **Workspace** tab), indicated by a green dot on the file's icon.

> 

Reviewer 2: We might also want to mention the fact that many developers actually don't write HTML directly, ever. They use server-side languages (Java, Python, PHP), or site generators (Eleventy), or even client-side frameworks that use JavaScript code, and other markup languages than HTML.

So, in reality, very few devs will want (or have the opportunity) to change HTML files in the Workspace tab. They will most likely edit another souce file (maybe a React component JS source file), then rebuild, then reload (both of which may happen automatically for them).

Author: ```
   Many developers don't write HTML directly, but instead use approaches such as:
   * A server-side language, such as Java, Python, or PHP.
   * A site generator, such as Eleventy.
   * A client-side framework that uses JavaScript code.
   * A markup language other than HTML, such as Markdown.

   This means that in practice, few developers change HTML files in the **Workspace** 
tab.  Most devs produce HTML files by using a different type of source file, such as a React 
component JS source file, and then rebuild and reload (which may happen automatically).
```

Continue with the next section.

<!-- ====================================================================== -->
## Step 4: Save a JavaScript change to disk
<!-- https://developer.chrome.com/docs/devtools/workspaces/#js -->
<!--
subsection order for Edge doc:
1. Open the Quick source tool without using Command Menu.
2. Open the Quick source tool by using Command Menu.
3. Open a .js file in the Quick source tool without using Command Menu.
4. Open a .js file in the Quick source tool by using Command Menu.
-->

Next, edit JavaScript and save changes to the source file.

The main place to use the code editor of DevTools is the **Sources** tool.  But sometimes you need to access other tools, such as the **Elements** tool or the **Console**, while editing files.  The **Quick source** tool gives you just the editor from the **Sources** tool, while any tool is open.

> There are a couple ways to open the **Quick source** tool, and a couple ways to open a file in the **Quick source** tool.  We'll walk through the regular, GUI way and then the Command Menu shortcut way.

Reviewer 2: I'm not convinced that we should talk about the Quick source tool here. We have other articles that mention Quick source and explain how to open it and make it display files. This article is all about the Workspace tab. The Workspace tab is already really complicated as it is. I'm not sure we're helping much by telling people that they can also change code in the Quick source tool.

I think we should focus this article on the main scenario for using Workspace, which is to set it up and then use it in the Sources tool.
And then, if people really want to view and edit source files at the same time as using another tool, they can go and learn about it in the article that's about Quick source.

Author: I was wondering/ doubting the same thing: seems like a jump of topic, to try to make Workspace tut teach instead about Quick source tool. @captainbrosset

Author: removed QS, redid png

Continue with the next section.

<!-- ====================================================================== -->
## Step 4: Save a JavaScript change to disk
<!-- https://developer.chrome.com/docs/devtools/workspaces/#js -->
<!--
subsection order for Edge doc:
1. Open the Quick source tool without using Command Menu.
2. Open the Quick source tool by using Command Menu.
3. Open a .js file in the Quick source tool without using Command Menu.
4. Open a .js file in the Quick source tool by using Command Menu.
-->

Next, edit JavaScript and save changes to the source file.

The main place to use the code editor of DevTools is the **Sources** tool.  But sometimes you need to access other tools, such as the **Elements** tool or the **Console**, while editing files.  The **Quick source** tool gives you just the editor from the **Sources** tool, while any tool is open.

There are a couple ways to open the **Quick source** tool, and a couple ways to open a file in the **Quick source** tool.  We'll walk through all the ways.

<!-- ------------------------------ -->

> #### Open the Quick source tool without using Command Menu

Author: maybe a better, tighter sequence than
![image](https://github.com/user-attachments/assets/611d2424-4f8c-4653-a374-7b222645a681)
would be two h4s:

Open Quick Source tool & a .js file [w/o using Command Menu]
Open Quick Source tool & a .js file by using Command Menu

@captainbrosset

Author: done.  it's better, and matches closely w/ upstream outline
